erasure-coding/src/lib.rs

```diff
@@ -0,0 +1,409 @@
+// This file is part of Polkadot.
+
+// Polkadot is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Polkadot is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+//! As part of Polkadot's availability system, certain pieces of data
+//! for each block are required to be kept available.
+//!
+//! The way we accomplish this is by erasure coding the data into n pieces
+//! and constructing a merkle root of the data.
+//!
+//! Each of n validators stores their piece of data. We assume n=3f+k, k < 3.
+//! f is the maximum number of faulty vaildators in the system.
+//! The data is coded so any f+1 chunks can be used to reconstruct the full data.
+
+extern crate polkadot_primitives as primitives;
+extern crate reed_solomon_erasure as reed_solomon;
+extern crate parity_codec as codec;
+extern crate substrate_primitives;
+extern crate substrate_trie as trie;
+
+use codec::{Encode, Decode};
+use reed_solomon::galois_16::{self, ReedSolomon};
+use primitives::{Hash as H256, BlakeTwo256, HashT};
+use primitives::parachain::{BlockData, Extrinsic};
+use substrate_primitives::Blake2Hasher;
+use trie::{MemoryDB, Trie, TrieMut, TrieDB, TrieDBMut};
+
+use self::wrapped_shard::WrappedShard;
+
+mod wrapped_shard;
+
+// we are limited to the field order of GF(2^16), which is 65536
+const MAX_VALIDATORS: usize = <galois_16::Field as reed_solomon::Field>::ORDER;
+
+/// Errors in erasure coding.
+#[derive(Debug, Clone)]
+pub enum Error {
+	/// Returned when there are too many validators.
+	TooManyValidators,
+	/// Cannot encode something for no validators
+	EmptyValidators,
+	/// Cannot reconstruct: wrong number of validators.
+	WrongValidatorCount,
+	/// Not enough chunks present.
+	NotEnoughChunks,
+	/// Too many chunks present.
+	TooManyChunks,
+	/// Chunks not of uniform length or the chunks are empty.
+	NonUniformChunks,
+	/// An uneven byte-length of a shard is not valid for GF(2^16) encoding.
+	UnevenLength,
+	/// Chunk index out of bounds.
+	ChunkIndexOutOfBounds(usize, usize),
+	/// Bad payload in reconstructed bytes.
+	BadPayload,
+	/// Invalid branch proof.
+	InvalidBranchProof,
+	/// Branch out of bounds.
+	BranchOutOfBounds,
+}
+
+struct CodeParams {
+	data_shards: usize,
+	parity_shards: usize,
+}
+
+impl CodeParams {
+	// the shard length needed for a payload with initial size `base_len`.
+	fn shard_len(&self, base_len: usize) -> usize {
+		(base_len / self.data_shards) + (base_len % self.data_shards)
```

Comment:
I'd expect `(base_len-1) / self.data_shards + 1` here.

Reply:
After the confusion below I'd expect `(base_len-1+8) / self.data_shards + 1` here.  And modify the last shard to state its length at the end.  And @drskalman points out that parity codec just ads the length too.

Reply:
8 seems very arbitrary?

Reply:
`base_len / self.data_shards + (base_len % self.data_shards == 0) as usize`